### PR TITLE
Autoconversion port to C++/Kokkos. 

### PR DESCRIPTION
**Pull request by pressel**
This PR is a port of P3's cloud_water_autoconversion subroutine from Fortran into C++/Kokkos. The port currently lacks a physics unit tests and has only been tested in double precision on CPUs.  Calling the C++ code from Fortran is B4B and the code passes unit B4B tests. 

This PR is no longer BFB as BFB math macros were added to micro_p3.F90 in order to ensure BFB between Fortran and C++.

### CODE REVIEW

File: components/scream/src/physics/p3/tests/p3_unit_tests.cpp [hunk 1]
```diff
@@ -91,4 +201,9 @@ TEST_CASE("p3_functions", "[p3_functions]")
   scream::p3::unit_test::UnitWrap::UnitTest<scream::DefaultDevice>::TestP3Func::run();
 }
 
+TEST_CASE("p3_cloud_water_autoconversion_test", "[p3_cloud_water_autoconversion_test]"){
+  //scream::p3::unit_test::UnitWrap::UnitTest<scream::DefaultDevice>::TestP3Conservation::run();
```
Comment (pressel) [inline components/scream/src/physics/p3/tests/p3_unit_tests.cpp]: This commented out line is where the physics unit tests will be called.

File: components/scream/src/physics/p3/p3_functions_autoconversion_impl.hpp [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    if(qc_not_small.any()){
+        qcaut.set(qc_not_small,
+            sp(1350.0)*pow(qc_incld,sp(2.47))*pow(nc_incld*sp(1.e-6)*rho,sp(-1.79)));
+        // note: ncautr is change in Nr; ncautc is change in Nc
+        ncautr.set(qc_not_small, qcaut*C::CONS3);
+        ncautc.set(qc_not_small, qcaut*nc_incld/qc_incld);
+    }
+
+    const auto qcaut_is_zero = qcaut == 0;
+    if(qcaut_is_zero.any()){ncautc.set(qcaut_is_zero,  0.0);}
```
Comment (ambrad) [inline components/scream/src/physics/p3/p3_functions_autoconversion_impl.hpp]: For one line computations, don't call `any()` first; just do the computation with the masked set. In this case, you're just setting masked values to 0, so there's even more reason to avoid the unnecessary conditional.

File: components/scream/src/physics/p3/p3_functions.hpp [hunk 3]
```diff
@@ -246,21 +246,24 @@ struct Functions
...
   static void get_cloud_dsd2(
     const Smask& qc_gt_small, const Spack& qc, Spack& nc, Spack& mu_c, const Spack& rho, Spack& nu,
     const view_1d<const Scalar>& dnu, Spack& lamc, Spack& cdist, Spack& cdist1, const Spack& lcldm);
 
   // Computes and returns rain size distribution parameters
-  KOKKOS_INLINE_FUNCTION
+  KOKKOS_FUNCTION
   static void get_rain_dsd2 (
     const Smask& qr_gt_small, const Spack& qr, Spack& nr, Spack& mu_r,
     Spack& lamr, Spack& cdistr, Spack& logn0r, const Spack& rcldm);
```
Comment (pressel) [inline components/scream/src/physics/p3/p3_functions.hpp]: Switching from KOKKOS_INLINE_FUNCTION to KOKKOS_FUNCTION fixed a warning thrown by GCC.

File: components/scream/src/physics/p3/p3_functions_autoconversion_impl.hpp (hunk 2 above)
Comment (pressel) [reply components/scream/src/physics/p3/p3_functions_autoconversion_impl.hpp]: Good point. Thanks

File: components/scream/src/physics/p3/p3_functions_autoconversion_impl.hpp [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+#include "p3_constants.hpp"
+
+namespace scream {
+namespace p3 {
+
+template<typename S, typename D>
+KOKKOS_FUNCTION
+void Functions<S,D>
+::cloud_water_autoconversion(const Spack& rho, const Spack& inv_rho, const Spack& qc_incld, const Spack& nc_incld,
+    const Spack& qr_incld, const Spack& mu_c, const Spack& nu, Spack& qcaut, Spack& ncautc, Spack& ncautr)
```
Comment (pressel) [inline components/scream/src/physics/p3/p3_functions_autoconversion_impl.hpp]: Some of the variables are no longer needed to be passed, I will remove those.

File: components/scream/src/physics/p3/p3_functions.hpp [hunk 5]
```diff
@@ -246,21 +246,24 @@ struct Functions
   // Calls polysvp1 to obtain the saturation vapor pressure, and then computes
   // and returns the saturation mixing ratio, with respect to either liquid or ice,
   // depending on value of 'ice'
-  KOKKOS_INLINE_FUNCTION
+  KOKKOS_FUNCTION
```
Comment (jgfouca) [inline components/scream/src/physics/p3/p3_functions.hpp]: @ambrad , are there potential performance implications for de-inlining these functions?

Comment (ambrad) [reply components/scream/src/physics/p3/p3_functions.hpp]: Should be fine. We can de-inline a few small ones later during the optimization stage.

File: components/scream/src/physics/p3/p3_functions_f90.cpp [hunk 6]
```diff
@@ -289,6 +296,23 @@ void get_rain_dsd2_f(Real qr_, Real* nr_, Real* mu_r_, Real* lamr_, Real* cdistr
...
+    using P3F = Functions<Real, HostDevice>;
+
+    typename P3F::Spack rho(rho_);
+    typename P3F::Spack qc_incld(qc_incld_);
+    typename P3F::Spack nc_incld(nc_incld_);
+    typename P3F::Spack qcaut(*qcaut_);
+    typename P3F::Spack ncautc(*ncautc_);
+    typename P3F::Spack ncautr(*ncautr_);
+
+    P3F::cloud_water_autoconversion(rho, qc_incld, nc_incld, qcaut, ncautc, ncautr);
```
[the remark this answers is not in the document]
Comment (pressel) [reply components/scream/src/physics/p3/p3_functions_f90.cpp]: Sounds good to me, I was following what you had done in the earlier PR.

[the remark this answers is not in the document]
Comment (jgfouca) [reply components/scream/src/physics/p3/p3_functions_f90.cpp]: If only it were as easy as switching the template to DefaultDevice :)

Remember, unlike host, the device is only used from inside kernels, so you'll need to spawn a kernel before the call to cloud_water_autoconversion. If you need to see what this looks like, refer to recent master.

[the remark this answers is not in the document]
Comment (pressel) [reply components/scream/src/physics/p3/p3_functions_f90.cpp]: @jgfouca, I think it should now run on the device. I'm sure you will let me know if I have missed something.

[the remark this answers is not in the document]
Comment (jgfouca) [reply components/scream/src/physics/p3/p3_functions_f90.cpp]: @pressel , that looks better.

File: components/cam/src/physics/cam/micro_p3_utils.F90 [hunk 7]
```diff
@@ -1,7 +1,26 @@
+
+#ifdef SCREAM_CONFIG_IS_CMAKE
```
Comment (jgfouca) [inline components/cam/src/physics/cam/micro_p3_utils.F90]: I'm surprised you needed the bfb math layer for this file. We should probably move it to it's own header and #include it instead of duplicating it.

File: components/cam/src/physics/cam/micro_p3_utils.F90 [hunk 8]
```diff
@@ -195,7 +214,7 @@ subroutine micro_p3_utils_init(cpair,rair,rh2o,rhoh2o,mwh2o,mwdry,gravit,latvap,
 
     cons1 = piov6*rhow
     cons2 = 4._rtype*piov3*rhow
-    cons3 = 1._rtype/(cons2*(25.e-6_rtype)**3)
+    cons3 = 1._rtype/(cons2*1.562500000000000d-14)  ! 1._rtype/(cons2*bfb_pow(25.e-6_rtype,3.0_rtype))
```
Comment (jgfouca) [inline components/cam/src/physics/cam/micro_p3_utils.F90]: It looks like you aren't actually using the bfb math layer, so you shouldn't need the cpp stuff you added to the top of this file.

File: components/scream/scripts/const-maker [hunk 9]
```diff
@@ -42,7 +42,10 @@ def const_maker(varname, expr):
     # them double-precision. The assignment to real(rtype) will truncate if
     # necessary.
     print("Fortran:")
-    print("real(rtype), public, parameter :: {} = {:22.15d} ! {}".format(varname, val, expr))
+    s = "real(rtype), public, parameter :: {} = {:22.15e} ! {}".format(varname, val, expr)
+    s = s.replace('e+','d+')
```
Comment (jgfouca) [inline components/scream/scripts/const-maker]: why was this necessary?

File: components/scream/src/physics/p3/tests/p3_unit_tests.cpp [hunk 10]
```diff
@@ -80,6 +80,127 @@ struct UnitWrap::UnitTest<D>::TestP3Func
...
+{
+
+static void  cloud_water_autoconversion_unit_bfb_tests(){
+  using KTH = KokkosTypes<HostDevice>;
+
+  static constexpr Int max_pack_size = 16;
+  REQUIRE(Spack::n <= max_pack_size);
+
+  CloudWaterAutoconversionData cwadc[max_pack_size] = {
+    { 0.97026902585098274, 5.1000000000000004e-3, 206128398.07453227,  0.0, 0.0, 0.0},
```
Comment (jgfouca) [inline components/scream/src/physics/p3/tests/p3_unit_tests.cpp]: You don't need to explicitly initialize the outputs to 0.0, that's the default behavior.

File: components/cam/src/physics/cam/micro_p3_utils.F90 (hunk 7 above)
Comment (pressel) [reply components/cam/src/physics/cam/micro_p3_utils.F90]: That's correct the math layer is not needed here, I have thus removed it from this file.

File: components/scream/scripts/const-maker (hunk 9 above)
Comment (pressel) [reply components/scream/scripts/const-maker]: The "d" in ":22.15d" is correct for integers but not floats and thus was a bug in const-maker. @ambrad provided me with a patch to fix this bug and asked me to include it as part of this PR.

File: components/scream/src/physics/p3/tests/p3_unit_tests.cpp (hunk 10 above)
Comment (pressel) [reply components/scream/src/physics/p3/tests/p3_unit_tests.cpp]: I removed the explicit initialization of 0.0. Thanks for catching this!

**Comment by pressel** [comment]
This PR is now passing all tests that are expected to pass for both CPU and GPU on Summit. The PR should be ready to merge upon final code review.

**Comment by jgfouca** [comment]
Looks good!

**Comment by bartgol** [comment]
> This PR is now passing all tests that are expected to pass for both CPU and GPU on Summit. The PR should be ready to merge upon final code review.

I'm removing the WIP tag, and let the Autotester test the PR on the target archs on sandia machines.

**Comment by bartgol** [comment]
@pressel I see both a BFB and Non-BFB label. Which one is it?

**Comment by jgfouca** [comment]
@bartgol , I fixed it. This PR is not bfb due to use of the cxx math routines in fortran.

**Comment by pressel** [comment]
@bartgol, the test that is failing is p3_run_and_cmp_ut_np1_omp1.

**Comment by jgfouca** [comment]
Conflicts resolved, running a brief check and then I'll merge.
